File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    '''
    Morris Algorithm
    '''
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []
        else:
            inorder = []
            curr = root
            while curr:
                if curr.left is None:
                    inorder.append(curr.val)
                    curr = curr.right
                else:
                    prev = curr.left
                    while prev.right and prev.right != curr:
                        prev = prev.right
                    if prev.right is None:
                        prev.right = curr 
                        curr = curr.left
                    else :
                        prev.right = None
                        inorder.append(curr.val)
                        curr = curr.right
            return inorder

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 import java.util.*;
class Solution {

    public List<Integer> inorderTraversal(TreeNode root) {
    
        ArrayList<Integer> inorderList = new ArrayList<>();
        Stack<TreeNode> stk = new Stack<>();
        TreeNode curr = root;
        while (!stk.isEmpty() || curr != null)
        {   
            while (curr != null)
            {
                stk.push(curr);
                curr = curr.left;
            }
            curr = stk.pop();
            inorderList.add(curr.val);
            curr = curr.right;
        }
        return inorderList; 

    }
}